ap_src/ap_app/utils/switch_permissions.py

```diff
@@ -0,0 +1,68 @@
+# Models
+from ..models import (UserProfile)
+from django.contrib.auth.models import User
+
+# Utils
+from teams_utils import get_users_sharing_teams
+from user_profile import get_user_id_from_username
+from user_profile import get_userprofile_id_from_user_id
+
+# this check should be activated when the user leaves a team
+def check_for_lingering_switch_perms(username): # stops users from having switch perms when they don't share any teams
+    user_id = get_user_id_from_username(username)
+    userprofile_id = get_userprofile_id_from_user_id(user_id)
+
+    usernames_given_permissions, userprofile_usernames_who_give_permissions = get_associated_permissions(username, userprofile_id, user_id)
+
+    user = User.objects.get(id=user_id)
+    users_sharing_teams = get_users_sharing_teams(username, user)
+    
+    process_user_usernames(username, usernames_given_permissions, users_sharing_teams)
+    process_userprofile_usernames(username, userprofile_usernames_who_give_permissions, users_sharing_teams)
+    
+    """
+    SET user ID whose absences are being edited AS (int) user_ID_edited
+    SET user IDs with permission to edit absences AS (list of int) user_IDs_with_perms
+    SET user IDs who are in the same team as user_ID_edited AS (list of int) user_IDs_sharing_teams
+
+    IF user_ID_edited_leave_team THEN
+    FOR EACH (int) user_ID_with_perms FROM (list) user_IDs_with_perms DO
+        IF user_ID_with_perms NOT IN user_IDs_sharing_teams THEN remove_permissions
+    END FOREACH
+    END IF
+    """
+
+def get_associated_permissions(current_user, selected_userprofile_id, selected_user_id):
+    usernames_given_permissions = set(User.objects.filter(permissions=selected_userprofile_id).values_list("username", flat=True))
+    usernames_given_permissions.remove(current_user)
+
+    userprofile_ids_who_give_permissions = UserProfile.objects.filter(edit_whitelist=selected_user_id).values_list(flat=False)
+    userprofile_usernames_who_give_permissions = []
+    for userprofile_id in userprofile_ids_who_give_permissions:
+        user = User.objects.get(userprofile=userprofile_id)
+        current_username = user.get_username()
+        userprofile_usernames_who_give_permissions.append(current_username)
+    userprofile_usernames_who_give_permissions = set(userprofile_usernames_who_give_permissions)
```

Comment:
You should be able to simplify this block of code with a statement like this:

`userprofile_usernames_who_give_permissions = set(User.objects.filter(userprofile__in=userprofile_ids_who_give_permissions).values_list("username", flat=True))`

If that doesn't work though I wouldn't bother spending too much time trying to refactor it. There's usually more efficient ways of doing things, but sometimes if it works then that's good enough.

Reply:
You are right, the `for` loop is inefficient and will not scale well. However, I think this should be made a separate issue because excessive time would be spent testing it again to see if it works.

Reply:
Fixed with ccaa924